datalog_compiler/src/lib.rs

```diff
@@ -0,0 +1,545 @@
+                let entry = identifier_to_bindings
+                    .entry(param.clone())
+                    .or_insert_with(HashMap::new);
+                entry.insert(source_idx, i);
+            }
+        }
+
+        let mut sorted_identifiers = identifier_to_bindings.keys().cloned().collect::<Vec<_>>();
+        sorted_identifiers.sort_by_key(|ident| ident.name.clone());
+
+        let identifiers_to_join = sorted_identifiers
+            .iter()
+            .filter_map(|ident| {
+                let bindings = identifier_to_bindings.get(ident).unwrap();
+                if bindings.len() > 1 {
+                    Some((ident.clone(), bindings.clone()))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let join_node = syn::Ident::new(&format!("{}_join", target_ident), Span::call_site());
+
+        let output_data = rule
+            .target
+            .fields
+            .iter()
+            .map(|field| {
+                let bindings = identifier_to_bindings.get(field).unwrap();
+                let source = bindings.keys().min().unwrap();
+                let source_expr: syn::Expr = if sources.len() == 1 {
+                    parse_quote!(v)
+                } else {
+                    let pair_idx = syn::Index::from(*source);
+                    parse_quote!(kv.1.#pair_idx)
+                };
+
+                let source_col_idx = syn::Index::from(*bindings.get(source).unwrap());
+
+                parse_quote!(#source_expr.#source_col_idx)
+            })
+            .collect::<Vec<syn::Expr>>();
+
+        // TODO(shadaj): this is all a hack, will fall apart when we have more than usize types
+        let source_types = sources
+            .iter()
+            .map(|source| {
+                let col_types = source
+                    .fields
+                    .iter()
+                    .map(|_| parse_quote!(usize))
+                    .collect::<Vec<syn::Type>>();
+
+                parse_quote!((#(#col_types, )*))
+            })
+            .collect::<Vec<syn::Type>>();
+
+        let key_type = identifiers_to_join
+            .iter()
+            .map(|(_, _)| parse_quote!(usize))
+            .collect::<Vec<syn::Type>>();
+
+        let after_join_map: syn::Expr = if sources.len() == 1 {
+            parse_quote!(|v| (#(#output_data, )*))
+        } else {
+            parse_quote!(|kv: ((#(#key_type, )*), (#(#source_types, )*))| (#(#output_data, )*))
+        };
+
+        let merge_index = merge_counter.entry(target.name.clone()).or_insert(0);
+        let my_merge_index = *merge_index;
+        *merge_index += 1;
+
+        let after_join = Pipeline::Link(PipelineLink {
+            lhs: Box::new(Pipeline::Operator(Operator {
+                path: parse_quote!(map),
+                paren_token: Paren::default(),
+                args: vec![after_join_map].iter().cloned::<syn::Expr>().collect(),
```

Comment:
`Punctuated::from_iter([after_join_map.clone()])` (though whole thing could be one parse_quote!)